numpy/lib/_index_tricks_impl.py

```diff
@@ -682,16 +684,24 @@ class ndindex:
     (1, 1, 0)
     (2, 0, 0)
     (2, 1, 0)
+    
+    Notes
+    -----
+    As of NumPy 2.4.0.dev0, this iterator is implemented using `itertools.product`
+    from Python's standard library. This change provides significant improvements
+    in both performance and memory efficiency, particularly for large iteration
+    spaces, while maintaining the original behavior and interface.
 
     """
 
+
     def __init__(self, *shape):
         if len(shape) == 1 and isinstance(shape[0], tuple):
             shape = shape[0]
-        x = as_strided(_nx.zeros(1), shape=shape,
-                       strides=_nx.zeros_like(shape))
-        self._it = _nx.nditer(x, flags=['multi_index', 'zerosize_ok'],
-                              order='C')
+        self.shape = shape
+        self._iter = product(*(range(s) for s in shape))
```

Comment:
feel free to ignore this, but a `map` would also work here:

```suggestion
        self._iter = product(*map(range, shape))
```